### PR TITLE
refactor: Convert Maven-based datasources to class

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- `clojure`
- `maven`
- `sbt-package`
- `sbt-plugin`

## Context:

- Ref: #8647

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/maven/index.ts [hunk 1]
```diff
@@ -177,162 +80,286 @@ function extractSnapshotVersion(metadata: XmlDocument): string | null {
...
+    }
 
-  if (process.env.RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK) {
+    const versions = extractVersions(mavenMetadata);
+    const releaseMap = versions.reduce(
+      (acc, version) => ({ ...acc, [version]: null }),
+      {}
+    );
+    if (!authorization) {
+      await packageCache.set(cacheNamespace, cacheKey, releaseMap, 30);
```
Comment (zharinov) [inline lib/datasource/maven/index.ts:129]: This is why I didn't use decorator-based cache. Let's do this later maybe.

Comment (JamieMagee) [reply lib/datasource/maven/index.ts:129]: Is it possible to use the `cacheable` property of the decorator?

https://github.com/renovatebot/renovate/blob/e7a55d2d37485eeb8d5320c131be572d2cdaed28/lib/util/cache/package/decorator.ts#L79-L83

It's used to only cache the default registry in `crate`

https://github.com/renovatebot/renovate/blob/c1555efb3b6e273ee6771f7b9ca746f5dad4eab2/lib/datasource/crate/index.ts#L35-L36

Comment (zharinov) [reply lib/datasource/maven/index.ts:129]: I think it's possible, but decorator usage for Maven may require a pretty big refactoring, and I would like to do this as the separate step.

Comment (JamieMagee) [reply lib/datasource/maven/index.ts:129]: Sounds good. I'll mark this as resolved.

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**